include/ibamr/LevelSetUtilities.h

```diff
@@ -0,0 +1,435 @@
+     * interface. Depending upon the application, ncells is used to specify material
+     * properties like density or viscosity (via some mixture model) or to compute volume
+     * of the phase enclosed by the level set variable.
+     */
+    void setInterfaceHalfWidth(double ncells)
+    {
+        d_ncells = ncells;
+    } // setInterfaceHalfWidth
+
+    double getInterfaceHalfWidth() const
+    {
+        return d_ncells;
+    } // getInterfaceHalfWidth
+
+    SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> getAdvDiffHierarchyIntegrator() const
+    {
+        return d_adv_diff_integrator;
+    } // getAdvDiffHierarchyIntegrator
+
+    SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > getLevelSetVariable(int idx = 0) const
+    {
+        return d_ls_vars[idx];
+    } // getLSVariable
+
+    SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> d_adv_diff_integrator;
+    std::vector<SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > > d_ls_vars;
+    double d_ncells = 1.0;
+    };
+
+    /*!
+     * \brief A lightweight class to tag grid cells containing the level set variable for grid refinement
+     */
+    class TagLSRefinementCells
+    {
+    public:
+        /*!
+         * \brief Constructor of the class.
+         */
+        TagLSRefinementCells(SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> adv_diff_integrator,
+                             SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_var,
+                             double tag_min_value = 0.0,
+                             double tag_max_value = 0.0)
+            : d_ls_container(adv_diff_integrator, ls_var),
+              d_tag_min_value(tag_min_value),
+              d_tag_max_value(tag_max_value)
+        {
+            // intentionally left blank
+        } // TagLevelSetRefinementCells
+
+        void setTagMinValue(double tag_min)
+        {
+            d_tag_min_value = tag_min;
+            return;
+        } // setTagMinValue
+
+        double getTagMinValue() const
+        {
+            return d_tag_min_value;
+        } // getTagMinValue
+
+        void setTagMaxValue(double tag_max)
+        {
+            d_tag_max_value = tag_max;
+            return;
+        } // setTagMaxValue
+
+        double getTagMaxValue() const
+        {
+            return d_tag_max_value;
+        } // getTagMaxValue
+
+        LevelSetContainer d_ls_container;
+
+    private:
+        double d_tag_min_value = 0.0;
+        double d_tag_max_value = 0.0;
+    }; // TagLevelSetRefinementCells
+
+    void tagLSCells(SAMRAI::tbox::Pointer<SAMRAI::hier::BasePatchHierarchy<NDIM> > hierarchy,
+                    const int level_number,
+                    const double error_data_time,
+                    const int tag_index,
+                    const bool initial_time,
+                    const bool uses_richardson_extrapolation_too,
+                    void* ctx);
```

Comment:
Can you document what this function does? Note in the documentation that `ctx` must be of type `TagLevelSetRefinementCells`.

Reply:
These are required signatures for working with the callbacks in `HierarchyIntegrator`.

Reply:
Got it --- I'll remove those other comments.